sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslUdfUdafTest.java

```diff
@@ -66,6 +67,54 @@ public void testUdaf() throws Exception {
     pipeline.run().waitUntilFinish();
   }
 
+  /**
+   * Test that an indirect subclass of a {@link CombineFn} works as a UDAF.
+   * BEAM-3777
+   */
+  @Test
+  public void testUdafMultiLevelDescendent() {
+    RowType resultType = RowSqlType.builder()
+        .withIntegerField("f_int2")
+        .withIntegerField("squaresum")
+        .build();
+
+    Row row = Row.withRowType(resultType).addValues(0, 354).build();
+
+    String sql1 = "SELECT f_int2, double_square_sum(f_int) AS `squaresum`"
+        + " FROM PCOLLECTION GROUP BY f_int2";
+    PCollection<Row> result1 =
+        boundedInput1.apply(
+            "testUdaf",
+            BeamSql.query(sql1).registerUdaf("double_square_sum", new SquareSquareSum()));
+    PAssert.that(result1).containsInAnyOrder(row);
+
+    pipeline.run().waitUntilFinish();
+  }
+
+  /**
+   * Test that correct exception is thrown when subclass of {@link CombineFn} is not parameterized.
+   * BEAM-3777
+   */
+  @Test
+  public void testRawCombineFnSubclass() {
+    exceptions.expect(IllegalStateException.class);
```

Comment:
Can you also add a loose message check on the keywords? Like `exceptions.expectMessage("parameterized")` and `exceptions.expectMessage("CombineFn")`.

Not a huge deal, but makes sure other weird crashes don't cause this to pass.

Reply:
will do

Reply:
@kennknowles done